CASSANDRA-20086: Use score ordered iterators for ANN search

michaeljmarshall opened the pull request:
This PR implements the feature proposed in https://issues.apache.org/jira/browse/CASSANDRA-20086.

## Motivation

Current ANN search does the following:

1. Get top k vectors from each index segment (memory or sstable)
2. Sort them in Primary Key order
3. Consume and materialize all of of them (`LIMIT * numSegments`)
4. Re-query with higher LIMIT if shadow keys are encountered (updates/overwrites)
5. Re-score to sort by similarity score
6. Take LIMIT best vectors
7. Sort by Primary Key

This flow is both inefficient due to multiple sorts/requeries and can produce invalid results in cases where a row has an update to a vector to a worse score.

The new flow:

1. Get top k vectors from each index segment (memory or sstable)
2. Build score ordered iterators
3. Merge those iterators, sorting by score descending
4. Consume, materialize, and validate rows from the global iterator until we have LIMIT rows
5. If an iterator is exhausted during step 4, requery the exhausted iterator's graph to get more vectors, using a bitset to ignore already visited nodes, until the graph is exhausted.
6. Sort LIMIT rows by primary key.

This change removes two sort operations. It also fixes all update cases by asserting that a row's position within the score ordered iterator is only valid if the vector cell's materialized value is from the same sstable.

## Other minor changes

1.  Increased test coverage by parameterizing the logic to test brute force, graph search, and "graph's choice".
2. Implemented custom brute force iterator that utilizes two priority queues to do the PQ based ranking then the full resolution ranking
3. Fixed a bug in the mapping of `limit` to `topk` where we could get `topk < limit` for limit 1000

I plan on completing some basic bench marks next week to demonstrate the value of this change. For now, I am opening this to start to get some reviews.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexQueryPlan.java:
```diff
@@ -150,7 +150,7 @@ public Function<PartitionIterator, PartitionIterator> postProcessor(ReadCommand
 
         // in case of top-k query, filter out rows that are not actually global top-K
-        return partitions -> (PartitionIterator) new VectorTopKProcessor(command).filter(partitions);
+        return partitions -> (PartitionIterator) new VectorTopKProcessor(command).consumeSortByScoreAndTakeTopK(partitions);
```
So my understanding here is that, because replicas are still returning results in primary key order, we should have no cross-version compatibility issues? (i.e. `consumeSortByScoreAndTakeTopK()` has been renamed, but not actually modified.)

michaeljmarshall replied:
Correct, the a peer will not observe any change in behavior

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -120,24 +132,45 @@ public PartitionIterator filterReplicaFilteringProtection(PartitionIterator full
-                long lastShadowedKeysCount = queryContext.vectorContext().getShadowedPrimaryKeys().size();
-                ResultRetriever result = resultSupplier.get();
-                UnfilteredPartitionIterator topK = (UnfilteredPartitionIterator) new VectorTopKProcessor(command).filter(result);
+                queryView.referencedIndexes.forEach(SSTableIndex::releaseQuietly);
```
So we have materialized all of the data we need from the referenced SSTables by the time we return from `takeTopKThenSortByPrimaryKey()` correct?

maedhroz replied:
The only thing I'd say is that it might be good to at least mention that we rely on `takeTopKThenSortByPrimaryKey()` eagerly consuming the data.

maedhroz replied:
Also, I wonder if we could just move the release logic into `QueryView` itself.

michaeljmarshall replied:
I'll add a comment and make the `QueryView` implement `AutoCloseable` so we can just do try-with-resources here.

michaeljmarshall replied:
This is implemented now.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java:
```diff
@@ -126,7 +127,10 @@ public PrimaryKeyMap.Factory newPrimaryKeyMapFactory(SSTableReader sstable)
     {
-        return version.onDiskFormat().newSSTableIndex(sstableContext, index);
+        if (isIndexEmpty(index.termType(), index.identifier()))
+            return index.termType().isVector() ? new EmptyIndex(sstableContext, index) : null;
```
Does returning `null` here subtly change what happens in the non-vector case?

michaeljmarshall replied:
From one perspective, returning `null` here maintains the current behavior for non-vector indexes. It's the returning of the `EmptyIndex` in the vector case that changes the behavior. This method is called when we update the sstables in the `IndexViewManager`. We use the result from this method to get a reference to each of the index's sstables when building the `View`. Returning `null` lets the view manager skip the sstable reference. If we return `null` here for vector, there are tests that fail when we have an sstable with no indexed vectors that contains a tombstone for the row. See `rangeRestrictedTestWithDuplicateVectorsAndAddNullVector`.

The test fails because in the vector query path, we use the `View` to pick the memtables and sstables we'll search when materializing rows so that we can validly call `isIndexDataValid` which relies on stable sstable id and memtable references for validation. If we exclude the empty index, we materialize the row incorrectly and produce invalid results. In the non-vector case, we actually acquire a new view of the data for every time we go to materialize the row from storage, avoiding this case entirely.

We could consider using the non-vector query views in the same way we use them for vector queries, but then we need to build logic to merge views in the case of more than one query predicate. We could also move the skip logic to a higher level to make the design more obvious within the `IndexViewManager`. I don't have an answer yet. I'll continue to think on it.

michaeljmarshall replied:
I pushed a commit to remove the branching. I added a detailed commit message explaining the choice. tl;dr: removing the branching means we have slightly higher memory usage, but not by much.

maedhroz commented on src/java/org/apache/cassandra/index/sai/view/IndexViewManager.java:
```diff
@@ -186,6 +179,13 @@ private Pair<Collection<SSTableIndex>, Collection<SSTableContext>> getBuiltIndex
+                    logger.debug(index.identifier().logMessage("No on-disk index was built for SSTable {} because the SSTable " +
+                                                               "had no indexable rows for the index."), sstableContext.descriptor());
+                    continue;
+                }
```
So by moving this here, we're now at least attempting to validate even when we know the index is empty, and we're implicitly using the `null` return value to identify the non-vector case. I don't *think* this will blow up validation above, as it detects empty indexes too...but I wonder if there's a way to cut down on the duplication...

michaeljmarshall replied:
I think this concern was taken care of via this commit https://github.com/michaeljmarshall/cassandra/commit/7e550ccfd1a330bd15e43a941bdbccbf3bb2c52f (which is now squashed but is preserved on my "pre-squash" branch in my own fork). Please resolve if you agree.

Note that we ran into several edge cases where we ultimately filter out these empty index objects, but for now, we keep them in the `view` so that the view can be considered complete.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/EmptyIndex.java:
```diff
@@ -0,0 +1,121 @@
+    @Override
+    public ByteBuffer maxTerm()
+    {
+        return null;
```
nit: `minTerm` and `maxTerm` are interesting. I guess we could make an argument to have them `throw new UnsupportedOperationException()`, since they should never be called from the range term tree builder. `toString()` might be an issue though.

michaeljmarshall replied:
Here are the current precedents. When flushing vector indexes from the memtable, we create the `SegmentMetadata` with `ByteBufferUtil.bytes(0)` for the min and max terms. When building via the `SegmentBuilder` with compaction, we don't special case vector and therefore write the actual min or max vector according to the byte comparable representation. When we have an empty index, we don't write the `SegmentMetadata`.

I agree we cannot add these `EmptyIndex` objects to the `RangeTermTree`, but I tend to prefer `null` over the exceptional case because we already have some handling showing some assumption that the `minTerm()` and `maxTerm()` methods could return `null`. I don't feel too strongly though.

maedhroz replied:
Yup. I'll leave it to you.

michaeljmarshall replied:
Great, I am going to keep it as is.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/IndexSegmentSearcher.java:
```diff
@@ -84,6 +88,20 @@ else if (index.termType().isLiteral())
+     * @param context  to track per sstable cache and per query metrics
+     * @return an iterator of {@link PrimaryKeyWithScore} in descending score order
+     */
+    public CloseableIterator<PrimaryKeyWithScore> orderBy(Expression exp, AbstractBounds<PartitionPosition> keyRange, QueryContext context) throws IOException
```
nit: I'm not sure how much we want to bike-shed the naming of things. `ANN` is technically an `Expression`, but not in the sense that the other filters are. An "ordering" is an expression + direction, so this is almost that. Perhaps the direction is meaningless because nobody ever wants to get the "worst" scores. There's no directionality. Anyway, I'd be totally fine to call `exp` something like `ordering` if we feel like that helps communicate something about what we're doing here at the method signature level.

michaeljmarshall replied:
I am happy to change it to `ordering`. I put it as `exp` because it is, strictly speaking, an expression. However, if you think it is more readable the other way, I'll make the switch.

maedhroz replied:
Yeah, I guess the best we can do for readability is name the thing what it actually is regardless of the type and go from there.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/IndexSegmentSearcher.java:
```diff
@@ -84,6 +88,20 @@ else if (index.termType().isLiteral())
+     * @param exp      the object containing the ordering logic
+     * @param keyRange key range specific in read command, used by ANN index
+     * @param context  to track per sstable cache and per query metrics
+     * @return an iterator of {@link PrimaryKeyWithScore} in descending score order
```
nit: I'd throw this on a new line. Just helps separate it visually from the params.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java:
```diff
@@ -112,10 +109,23 @@ public KeyRangeIterator search(Expression expression, AbstractBounds<PartitionPo
+
     @Override
-    public KeyRangeIterator limitToTopKResults(QueryContext context, List<PrimaryKey> primaryKeys, Expression expression) throws IOException
+    public CloseableIterator<PrimaryKeyWithScore> orderResultsBy(QueryContext context, List<PrimaryKey> primaryKeys, Expression expression) throws IOException
```
nit: Same thing here. We could rename `expression` -> `ordering` if you think that communicates more about what's going on.

michaeljmarshall replied:
I went with `orderer` for the name, let me know if you prefer `ordering`.

maedhroz commented on src/java/org/apache/cassandra/index/sai/iterators/PriorityQueueIterator.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * An iterator over a priority queue.
+ * @param <T> the type of the elements in the priority queue
+ */
```
nit: It might make sense to indicate in the inline docs here that this iterator essentially consumes/destroys the PQ via `poll()` (even though that's a pretty reasonable assumption for a user to make).

michaeljmarshall replied:
I added a javadoc to the constructor indicating as much.

maedhroz commented on src/java/org/apache/cassandra/index/sai/memory/MemtableIndexManager.java:
```diff
@@ -92,6 +102,7 @@ public long update(DecoratedKey key, Row oldRow, Row newRow, Memtable memtable)
             return index(key, newRow, memtable);
         }
 
+        // TODO when do we hit this case? It isn't covered by tests.
```
Indexing a vector concurrently with a table or keyspace drop/truncation?

maedhroz replied:
Either way, this is a little unsettling, as the non-vector case always eventually produces a `MemtableIndex` to receive the write.

michaeljmarshall replied:
I would expect that we're guaranteed to not hit this case by the design of the update execution path but I don't know enough about the calling code to say for sure. However, with this change, we will have incorrect validation semantics if that assumption isn't true, so I think it might be better to add an assertion here that the value isn't `null`.

I agree that the branching isn't ideal, but note that we don't actually update anything in the non-vector case, we just insert but in the vector case, we're looking to replace the vector in the index.

Maybe it'd be helpful to get confirmation that the system is designed the way I think it is.

maedhroz replied:
Agreed an assertion is probably best. The only other thing I'd consider is just calling `maybeInitializeMemtableIndex(Memtable memtable)` -> `switchMemtable(Memtable next)` to follow the pattern of `renewMemtable()` and `discardMemtable()`

maedhroz commented on src/java/org/apache/cassandra/index/sai/utils/CellWithSourceTable.java:
```diff
@@ -0,0 +1,244 @@
+    private static final long EMPTY_SIZE = ObjectSizes.measure(new CellWithSourceTable<>(null, null, null));
+
+    private final Cell<T> cell;
+    private final Object sourceTable;
```
So here and in `RowWithSourceTable`, the "source" can be a Memtable or an SSTable ID? If we keep all this, wondering if that's an assumption we can codify on creation...

michaeljmarshall replied:
We do codify it on creation by offering two constructors taking either `Memtable` or `SSTableId`.

maedhroz replied:
Indeed. I read that bit after writing this comment. Apologies.

maedhroz commented on src/java/org/apache/cassandra/utils/CloseableIterator.java:
```diff
@@ -17,14 +17,29 @@
+    static <T> CloseableIterator<T> emptyIterator()
+    {
+        return (CloseableIterator<T>) EMPTY;
+    }
```
Is this functionally equivalent to `empty()` below...just optimized to avoid creating a new instance of `CloseableIterator`?

michaeljmarshall replied:
Yes, but I think we should remove this method. It isn't used on the typical hot path.

maedhroz commented on src/java/org/apache/cassandra/utils/CloseableIterator.java:
```diff
@@ -67,4 +82,35 @@ public T next()
+     * Pairs a {@link CloseableIterator} and an {@link AutoCloseable} so that the latter is closed when the former is
+     * closed.
+     */
+    static <T> CloseableIterator<T> withOnClose(CloseableIterator<T> iterator, Closeable onClose)
```
nit: unused method?

michaeljmarshall replied:
Removed it

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/PrimaryKeyWithScore.java:
```diff
@@ -0,0 +1,91 @@
+package org.apache.cassandra.index.sai.disk.v1.vector;
+
+import org.apache.cassandra.db.memtable.Memtable;
+import org.apache.cassandra.db.rows.Cell;
```
nit: unused import?

maedhroz commented on src/java/org/apache/cassandra/index/sai/memory/MemtableOrdering.java:
```diff
@@ -20,23 +20,36 @@
+     *
+     * Assumes that the given  spans the same rows as the implementing index's segment.
+     */
+    CloseableIterator<PrimaryKeyWithScore> orderResultsBy(QueryContext context, List<PrimaryKey> primaryKeys, Expression expression);
```
For this and `SegmentOrdering`, I've been mulling over naming a bit. You could imagine a signature like the one below, but take it as brainstorming rather than a strong suggestion. (I'm always trying to pack some meaning into argument names, so something like `primaryKeys` -> `results` is appealing to me.)

```
CloseableIterator<PrimaryKeyWithScore> orderResultsBy(QueryContext context, List<PrimaryKey> results, Expression scorer);
```

michaeljmarshall replied:
I made this change throughout the code base to all `orderResultsBy` methods.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/QueryController.java:
```diff
@@ -171,6 +175,39 @@ public UnfilteredRowIterator queryStorage(List<PrimaryKey> keys, ReadExecutionCo
+        if (key == null)
+            throw new IllegalArgumentException("non-null key required");
+
+        // TODO how do we want to handle static rows?
```
Looking at `readAndValidatePartition()`, the caller, are you worried about static rows not being wrapped with a `RowWithSourceTable` and breaking `isIndexDataValid()`? How would we test that...a query that only reads and filters static columns? ANN query against a single static column that forces us to check the validity of the static row?

michaeljmarshall replied:
Looks like this PR fails if the indexed column is static. Here is a minimal reproducer:

```java
    @Test
    public void testStaticVectorColumnIndex() throws Throwable
    {
        createTable("CREATE TABLE %s (pk int, ck int, val vector<float, 2> static, PRIMARY KEY(pk, ck))");
        createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex'");

        execute("INSERT INTO %s (pk, ck, val) VALUES (0, 1, [1,0])");
        execute("INSERT INTO %s (pk, ck)      VALUES (0, 2)");
        execute("INSERT INTO %s (pk, ck, val) VALUES (1, 3, [0,1])");

        beforeAndAfterFlush(() -> {
            Object[][] rows = rows(row(1), row(2));
            assertRows(execute("SELECT ck FROM %s ORDER BY val ANN OF [1,0] LIMIT 2"), rows);
        });
    }
```

The failure comes from this assertion in the calling code:

```
assert !clusters.hasNext() : "Expected only one row per partition";
```

If I remove that code, I discover that the `isIndexDataValid` seems incorrect here:

```java
        // If the row is static and the column is not static, or vice versa, the indexed value won't be present so we
        // don't need to check if live data matches indexed data.
        if (row.isStatic() != columnMetadata.isStatic())
            return true;
```

Next, I see that we have an issue in that we don't apply the row transformer to the `StaticRow` and therefore we cannot use `CellWithSource`.

I am considering two options:
1. Find a way to wrap rows and then cells (might be unnecessarily invasive)
2. Add logic to compute the vector similarity score and then we can consider a row valid if when the score is the same or better.

michaeljmarshall replied:
Looks like the `Transformation` class already has an `applyToStatic` method, so there is a path forward there.

michaeljmarshall replied:
this should be correct now. Resolving the comment, please reopen if you have additional thoughts.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/QueryViewBuilder.java:
```diff
@@ -57,12 +59,19 @@ public static class QueryExpressionView
+            // so it is safe to build a view fragment.
+            List<Memtable> memtables = memtableIndexes.stream().map(MemtableIndex::getMemtable).collect(Collectors.toList());
+            List<SSTableReader> sstableReaders = sstableIndexes.stream().map(SSTableIndex::getSSTable).collect(Collectors.toList());
+            this.viewFragment = new ColumnFamilyStore.ViewFragment(sstableReaders, memtables);
```
Not sure how we want to resolve yet, but just making a note that this adds some object creation overhead to the non-vector query path...I think...

michaeljmarshall replied:
I made it lazy

maedhroz commented on src/java/org/apache/cassandra/index/sai/utils/CellWithSource.java:
```diff
@@ -0,0 +1,244 @@
+
+    private CellWithSourceTable(ColumnMetadata column, Cell<T> cell, Object sourceTable)
+    {
+        super(column);
```
So I guess we could have something like...

```
assert sourceTable instanceof Memtable || sourceTable instanceof SSTableId || sourceTable == null : "Expected Memtable or SSTableId, got " + source;
```

michaeljmarshall replied:
We make this assertion in `RowWithSourceTable`. Do you want it here too? Seems reasonable, but is technically already covered.

maedhroz replied:
Yeah, it's probably fine if we're always using them together anyway.

maedhroz commented on src/java/org/apache/cassandra/index/sai/utils/CellWithSourceTable.java:
```diff
@@ -0,0 +1,244 @@
+    private final Cell<T> cell;
+    private final Object sourceTable;
+
+    public CellWithSourceTable(Cell<T> cell, Object sourceTable)
```
nit: I think `sourceTable` is just `source` in `RowWithSourceTable`. I wouldn't be opposed to simplifying at the class name level either, i.e. `CellWithSource` and `RowWithSource`

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/QueryController.java:
```diff
@@ -171,6 +175,39 @@ public UnfilteredRowIterator queryStorage(List<PrimaryKey> keys, ReadExecutionCo
+     * Get an iterator over the rows for this partition key. Restrict the search to the specified view.
+     * @param key
+     * @param executionController
+     * @return
```
nit: Incomplete JavaDoc elements. (I don't really care if we just remove the `@param` and `@return` bits if we don't have anything interesting to say about them.)

maedhroz commented on src/java/org/apache/cassandra/index/sai/memory/VectorMemoryIndex.java:
```diff
@@ -241,6 +253,30 @@ private int maxBruteForceRows(int limit, int nPermittedOrdinals, int graphSize)
         return (int) max(limit, memoryToDiskFactor * expectedComparisons);
```
Does `VectorMemoryIndex` ever actually hit disk? It seems like the only time it touches disk is on Memtable flush. So if the non-brute force case isn't hitting disk, what is it doing? It seems like it's searching the in-memory graph index and doing a bunch of node/ordinal -> primary key conversions.

(`orderBy()` and `orderResultsBy()` both seem to return score sorted keys now, so it doesn't seem like our choice of brute force vs. non-brute force has an impact on the amount of row materialization we do downstream...or am I reading that wrong?)

maedhroz replied:
...and if I'm correct about ^, the 0.25 is likely no longer useful.

michaeljmarshall replied:
The in memory vector index itself will not hit disk. Row materialization might, but this comment seems incorrect.

The decision point here is about the number of comparisons expected in graph traversal versus in brute force.

michaeljmarshall replied:
> (`orderBy()` and `orderResultsBy()` both seem to return score sorted keys now, so it doesn't seem like our choice of brute force vs. non-brute force has an impact on the amount of row materialization we do downstream...or am I reading that wrong?)

great observation

michaeljmarshall replied:
I switched the in memory and on disk logic for brute force. Take a look.

maedhroz commented on src/java/org/apache/cassandra/db/SinglePartitionReadCommand.java:
```diff
@@ -1008,8 +1033,9 @@ private UnfilteredRowIterator queryMemtableAndSSTablesInTimestampOrder(ColumnFam
-                result = add(RTBoundValidator.validate(iter, RTBoundValidator.Stage.SSTABLE, false),
+                UnfilteredRowIterator wrapped = rowTransformer != null ? Transformation.apply(iter, rowTransformer.apply(sstable.getId()))
+                                                                       : iter;
+                result = add(RTBoundValidator.validate(wrapped, RTBoundValidator.Stage.SSTABLE, false),
```
I don't think there's anything obviously unsafe here, but SPRC is a core part of the read path. I'll come back to it toward the end of review to double-check.

michaeljmarshall replied:
Any additional thoughts?

maedhroz replied:
Had a chance to look at this again, and the logic should be unchanged if the transformer is `null`. The only thing I think might be helpful is to have a marker interface, let's say we call it `RowSourceIdentifier` that `Memtable` and `SSTableId` can implement (perhaps w/ an `equals()` method in the interface just to make it clear that we need that, even though `Memtable` implementations all use reference equality).

michaeljmarshall replied:
I implemented this with the name `CellSourceIdentifier` since it ultimately only tells us a cell's source.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -245,35 +316,42 @@ public KeyRangeIterator limitToTopKResults(QueryContext context, List<PrimaryKey
+                                             : FORCE_BRUTE_FORCE_ANN;
     }
 
     private int maxBruteForceRows(int limit, int nPermittedOrdinals, int graphSize)
```
```
// brute force from sstable will also do a bunch of extra work (going through trie index to look up row).
```
I think this comment is no longer accurate w/ score-ordered results being returned from the searcher?

michaeljmarshall replied:
Correct, though I'm not sure when this changed. We rely on the vector graph having a copy of each row's vectors.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -62,22 +71,27 @@ public class VectorIndexSegmentSearcher extends IndexSegmentSearcher
 
+    // If true, use brute force. If false, use graph search. If null, use the normal logic.
+    @VisibleForTesting
+    public static Boolean FORCE_BRUTE_FORCE_ANN = null;
```
Worth throwing in `CassandraRelevantProperties`?

Something like `SAI_FORCE_BRUTE_FORCE_ANN("cassandra.sai.force_brute_force_ann", "false")`

michaeljmarshall replied:
I only added this to force tests to take the different paths. I don't think we should expose further unless you see a need.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/RowIdWithScore.java:
```diff
@@ -0,0 +1,48 @@
+        this.score = score;
+    }
+
+    public PrimaryKeyWithScore buildPrimaryKeyWithScore(ColumnMetadata columnMetadata,
```
```suggestion
    public PrimaryKeyWithScore toPrimaryKeyWithScore(ColumnMetadata columnMetadata,
```
...if you're a to/from fan :p

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -98,41 +118,44 @@ public KeyRangeIterator search(Expression exp, AbstractBounds<PartitionPosition>
+            {
+                int expectedNodesVisited = expectedNodesVisited(limit, graph.size(), graph.size());
+                IntConsumer nodesVisitedConsumer = nodesVisited -> updateExpectedNodes(nodesVisited, expectedNodesVisited);
+                return graph.search(queryVector, topK, limit, new Bits.MatchAllBits(graph.size()), nodesVisitedConsumer);
```
nit: This looks like the same code block as the one above on 136, but I'll let you decide whether you think factoring out a method makes sense.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -98,41 +118,44 @@ public KeyRangeIterator search(Expression exp, AbstractBounds<PartitionPosition>
     }
 
     /**
      * Return bit set we need to search the graph; otherwise return posting list to bypass the graph
```
nit: It's a private method, so I'd just remove the out-of-date comment altogether.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -141,20 +164,26 @@ private BitsOrPostingList bitsOrPostingListForKeyRange(VectorQueryContext contex
             // (nRows should not include shadowed rows, but context doesn't break those out by segment,
```
Are shadowed rows still a thing?

michaeljmarshall replied:
No, I'll remove the line

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -208,29 +236,72 @@ private SparseFixedBitSet bitSetForSearch()
     }
 
+    /**
+     * Produces a correct ranking of the rows in the given segment. Because this graph does not have compressed
```
Does the graph for sure not have vectors? If so, is the `if (graph.getCompressedVectors() != null)` check necessary on 250?

michaeljmarshall replied:
This comment is out of date. My first implementation was without the two pass logic, but now that we have it, we branch internally.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -208,29 +236,72 @@ private SparseFixedBitSet bitSetForSearch()
+        {
+            NeighborSimilarity.ExactScoreFunction esf = graph.getExactScoreFunction(queryVector, view);
+            NeighborQueue scoredRowIds = segmentOrdinalPairs.mapToSegmentRowIdScoreHeap(esf);
+            // TODO metrics? columnQueryMetrics.onBruteForceNodesReranked(segmentOrdinalPairs.size());
```
What would this metric tell an operator?

michaeljmarshall replied:
The metric could help expand what is happening at runtime, but there isn't much else. We have this same information in the trace at the moment, so I think we should defer on wiring up this metric.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/BruteForceRowIdIterator.java:
```diff
@@ -0,0 +1,117 @@
+ * <p>
+ * As an implementation detail, we use a heap to maintain state rather than a List and sorting.
+ */
+public class BruteForceRowIdIterator extends AbstractIterator<RowIdWithScore>
```
nit: One thing I like to do with classes that have some non-trivial state like this is throw the `@NotThreadSafe` annotation on the class (when it's clear they are meant to be used by a single thread).

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/BruteForceRowIdIterator.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    protected RowIdWithScore computeNext() {
```
nit: newline brace style

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/BruteForceRowIdIterator.java:
```diff
@@ -0,0 +1,117 @@
+        float score = exactScoreQueue.topScore();
+        int rowId = exactScoreQueue.pop();
+        return new RowIdWithScore(rowId, score);
+    }
```
So at the end of the day, the whole point of `BruteForceRowIdIterator` is to get an exact scoring/order without actually exactly scoring all of the vectors? (i.e. We get approximate results in batches and then calculate exact results on-demand.) I guess this is especially valuable when we have close to the limit for brute force in the first place? Would it be simpler just to lower the limit for brute force rather than having a hybrid like this?

michaeljmarshall replied:
The PQ vectors are in memory while the full precision ones are on disk. We want to get the top `rerankK` values based on the cheaper score and then we rerank based on the more expensive score. Brute force is needed when the result set is so restrictive that it is faster to just score all results instead of searching the graph and filtering out rows that don't satisfy the search predicate.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/DiskAnn.java:
```diff
@@ -86,13 +82,19 @@ public int size()
     {
         OnHeapGraph.validateIndexable(queryVector, similarityFunction);
 
+        // TODO why isn't this getting closed?
```
What would the best approach be? Closing the `GraphSearcher` as part of closing the `AutoResumingNodeScoreIterator`? Is there a test where this surfaces?

michaeljmarshall replied:
I didn't see this in a test. I just noticed it in passing and left myself a comment. I'm going to add logic to close it in `AutoResumingNodeScoreIterator`.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/QueryController.java:
```diff
@@ -171,6 +175,39 @@ public UnfilteredRowIterator queryStorage(List<PrimaryKey> keys, ReadExecutionCo
+                                                                                 makeFilter(List.of(key)));
+
+        // Class to transform the row to include its source table.
+        Function<Object, Transformation<BaseRowIterator<?>>> rowTransformer = (Object sourceTable) -> new Transformation<>()
```
I'm not sure how intelligent the compiler will be about extracting the `Transformation`. I guess we could make sure that's not an issue by just creating a constant and reusing it to avoid allocations?

michaeljmarshall replied:
Done.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -468,79 +501,321 @@ private UnfilteredRowIterator queryStorageAndFilter(List<PrimaryKey> keys)
-                    {
-                        matches.add(unfiltered);
-                        hasMatch = true;
+        // todo static keys??
```
Static-only rows are handled below on 543. Any particular concern here?

It looks like `filterPartition()` has been basically left intact for the non-vector case.

maedhroz replied:
Static rows have always been an extremely annoying thing to test with SAI queries, and I don't know if that will be much different here. If we do allow indexing static columns of the `vector` type, there are a couple things we need to think through on the testing front:

1.) What happens when we index a vector that is a static column and (as is usually the case), there are many rows in the associated partition? If that vector is in the top K, do we give the user all the rows in the partition? Do we fail the assertion `readAndValidatePartition()`?

2.) What happens if we have a non-static column vector indexed, but there's also a filter on at least one non-vector SAI-indexed column? How do we resolve the STATIC and WIDE (?) `PrimaryKey` instances together?

To be honest, I would have no issue at all with simply not allowing SAI indexes on static vector columns. It might not save us from #2 above, but there are so many corner cases with static columns, I don't know if it's worth supporting. (i.e. What data model have we seen where someone really needs to do this and they don't simply have a simple partition key-only primary key?)

michaeljmarshall replied:
I don't actually think that this part of the code has an issue with static keys. I was worried initially, but after closer review, this looks fine to me. At the very least, my new static vector column index tests are passing.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -120,24 +132,45 @@ public PartitionIterator filterReplicaFilteringProtection(PartitionIterator full
+        return new QueryViewBuilder(Collections.singleton(planExpression), queryController.mergeRange()).build();
     }
 
     private class ResultRetriever extends AbstractIterator<UnfilteredRowIterator> implements UnfilteredPartitionIterator
```
`private final boolean topK;`

nit: unused?

michaeljmarshall replied:
Removed it

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -468,79 +501,321 @@ private UnfilteredRowIterator queryStorageAndFilter(List<PrimaryKey> keys)
+        /**
+         * Determine if the key is in one of the queried key ranges. We do not iterate through results in
+         * {@link PrimaryKey} order, so we have to check each range.
+         * @param key
```
nit: `key` is kind of dangling here...I'm fine with just removing it in the face of a pretty helpful overall description.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -468,79 +501,321 @@ private UnfilteredRowIterator queryStorageAndFilter(List<PrimaryKey> keys)
+        private final FilterTree filterTree;
+        private final QueryController controller;
+        private final ReadExecutionController executionController;
+        private final QueryContext queryContext;
```
There are a few shared fields between the retrievers: `filterTree`, `queryContext`, `controller`, etc. I don't know if there's enough to factor something common out, given the only strict duplication looks like maybe the `metadata()` method.

michaeljmarshall replied:
Done

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -468,79 +501,321 @@ private UnfilteredRowIterator queryStorageAndFilter(List<PrimaryKey> keys)
+                        }
+                    }
+                }
+                return isRowValid ? new PrimaryKeyIterator(partition, staticRow, row)
```
So to ground myself a bit, is it correct to say that we never attempt to batch rows in the same partition? In data models without clusterings, I guess this doesn't matter...

michaeljmarshall replied:
Correct. We could technically batch within a partition for further optimization, but it would add complexity to this path. As it is, we get primary keys in an essentially random order, so there isn't an automatic expectation that keys in the same partition must be near each other in the score ordered iterator.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java:
```diff
@@ -468,79 +501,321 @@ private UnfilteredRowIterator queryStorageAndFilter(List<PrimaryKey> keys)
+            FileUtils.closeQuietly(scoredPrimaryKeyIterator);
+        }
+
+        public static class PrimaryKeyIterator extends AbstractUnfilteredRowIterator
```
```suggestion
        public static class SingleRowIterator extends AbstractUnfilteredRowIterator
```

michaeljmarshall replied:
I removed this class in the latest commits in favor of using `SinglePartitionIterator` since static vector columns can produce multiple rows at this point in the code.

maedhroz commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java:
```diff
@@ -173,6 +173,15 @@ public Index.Indexer indexerFor(Predicate<Index> indexSelector,
+                // TODO is this a good solution? I only need this because we do not always
+                //  indexed everything in insertRow and updateRow.
+                for (Index.Indexer indexer : indexers)
+                    indexer.begin();
```
Yeah, it would be really nice to avoid iterator creation overhead here on a hot path. What modifications to `VectorMemoryIndex` don't flow through `insertRow()` or `updateRow()`?

michaeljmarshall replied:
In an alternate version, I implemented these other methods:

```java
        @Override
        public void partitionDelete(DeletionTime deletionTime)
        {
            // Initialize the memtable index to ensure proper SAI views of the data
            indexContext.initializeMemtableIndex(mt);
        }

        @Override
        public void rangeTombstone(RangeTombstone tombstone)
        {
            // Initialize the memtable index to ensure proper SAI views of the data
            indexContext.initializeMemtableIndex(mt);
        }

        @Override
        public void removeRow(Row row)
        {
            // Initialize the memtable index to ensure proper SAI views of the data
            indexContext.initializeMemtableIndex(mt);
        }
```

the issue here is making sure that the `View` object has a reference to all relevant memtables. I don't remember why I didn't use this logic when I first wrote this patch in August, but I think that might be better since it covers the actual missing cases without adding overhead to the insert/update paths.

maedhroz replied:
SAI obviously doesn't index deletes, but what you're saying is that we need to do the initialization even when the only operations on the Memtable are deletes? Ideally, even if that's true, we'd also want to avoid the iterator creation on delete when there are no Vector indexes, but I want to understand the first thing better...

michaeljmarshall replied:
I updated the logic so we can use the `MemtableSwitchedNotification` notification and eagerly create the index object when it is a vector column.

michaeljmarshall replied:
Looks like I forgot to remove this code block, I'll remove in an upcoming commit.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -208,29 +236,72 @@ private SparseFixedBitSet bitSetForSearch()
+
+        // TODO add heuristic to determine if we should use one pass even though we have PQ
+        if (graph.getCompressedVectors() != null)
+            return orderByBruteForceTwoPass(graph.getCompressedVectors(), queryVector, segmentOrdinalPairs, limit, topK);
```
Does anything in the vector tests we modified in this patch test `orderByBruteForceTwoPass()`?

michaeljmarshall replied:
I added a test to the `VectorSiftSmallTest` to make sure we do.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/DiskAnn.java:
```diff
@@ -86,86 +83,56 @@ public int size()
-            scoreFunction = compressedVectors.approximateScoreFunctionFor(queryVector, similarityFunction);
-            reRanker = (i, map) -> similarityFunction.compare(queryVector, map.get(i));
+            FileUtils.closeQuietly(view);
+            throw new RuntimeException(e);
```
```suggestion
        throw Throwables.unchecked(e);
```

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/DiskAnn.java:
```diff
@@ -86,86 +83,56 @@ public int size()
+            return new NodeScoreToRowIdWithScoreIterator(nodeScoreIterator, ordinalsMap.getRowIdsView());
         }
-        else
+        catch (Exception e)
```
If we get an `Error` here like an `AssertionError`, do we properly close the view? Perhaps `Throwable` is safer?

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/Operation.java:
```diff
@@ -319,16 +320,23 @@ private static int getPriority(Operator op)
+            return controller.getTopKRows(view);
+
+        // Otherwise, we need to search first, then order.
         KeyRangeIterator iterator = Node.buildTree(controller.indexFilter()).analyzeTree(controller).rangeIterator(controller);
```
nit: Could we just call `buildIterator()` now instead of duplicating `Node.buildTree(controller.indexFilter()).analyzeTree(controller).rangeIterator(controller)`?

michaeljmarshall replied:
done.

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/QueryController.java:
```diff
@@ -313,117 +360,90 @@ public boolean doesNotSelect(PrimaryKey key)
+                    break;
+                primaryKeys.add(next);
+                if (MAX_MATERIALIZED_KEYS < ++count)
+                    throw new QueryMaterializesTooManyPrimaryKeysException("Too many primary keys. Attempted to load more than: " + MAX_MATERIALIZED_KEYS);
```
I've mulled this back and forth a bit, but I think we want to consider two things:

1.) Moving `max_materialized_keys` to something much less than 1M. `ReplicaFilteringProtectionOptions` is a similar place where we accumulate keys in memory, and it defaults to a warning at 2000 (and failure at 32k). I think we should go with something more conservative like 4096 (if we're into powers of 2).

2.) Rather than failing the query, we should just back out and hit the "orderBy" pathway if we hit the max/threshold. It's entirely possible that with an extremely non-selective filter, we'll still be able to complete the query before the timeout, because post-filtering won't filter many things out.

maedhroz commented on src/java/org/apache/cassandra/index/sai/memory/VectorMemoryIndex.java:
```diff
@@ -172,73 +186,92 @@ public KeyRangeIterator search(QueryContext queryContext, Expression expr, Abstr
-        double memoryToDiskFactor = 0.25;
-        return (int) max(limit, memoryToDiskFactor * expectedComparisons);
+        // ANN index will do a bunch of extra work besides the full comparisons (performing PQ similarity for each edge);
+        // VSTODO I'm not sure which one is more expensive (and it depends on things like sstable chunk cache hit ratio)
```
SSTable chunk cache has nothing to do with this right? At this point it's comparisons vs. in-memory graph node visits (and we're just kind of roughly treating them as being equal)?

michaeljmarshall replied:
The main question is the how many nodes we visit in the graph search. If it is more, then we should brute force it, if it is less, then we should search the graph. We can't know a priori though, so we have a heuristic.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
```diff
@@ -245,45 +311,54 @@ public KeyRangeIterator limitToTopKResults(QueryContext context, List<PrimaryKey
+        // in-memory comparisons are cheaper than pulling a row off disk and then comparing
+        // VSTODO this is dramatically oversimplified
+        // larger dimension should increase this, because comparisons are more expensive
+        // lower chunk cache hit ratio should decrease this, because loading rows is more expensive
```
Still trying to make sense of this...so the comparison here is between sorting already materialized `SegmentRowIdOrdinalPairs` and searching the on-disk graph? ...or does it still have to do with pulling rows?

michaeljmarshall replied:
IIUC, the comment is confusing because it applies to the old way of doing things. I'll update it and then you can check it. The real goal here is to minimize number of vector comparisons and graph nodes loaded from disk (these two are essentially all of the cost of vector search). If the permitted ordinals is small enough, then it's cheaper to pull the vectors from the disk, do the similarity score, and put the rows into a priority queue.

michaeljmarshall replied:
One of the key details is that we no longer read extra rows from disk because of the use of score ordered iterators.

michaeljmarshall commented:
I rebased/squashed to a single commit. I saved the original commits here: https://github.com/michaeljmarshall/cassandra/commits/cassandra-20086-pre-squash/.

maedhroz commented:
https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-5/417/

maedhroz commented:
Latest few commits look good...

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/VectorIndexSegmentSearcher.java:
looks like `updateExpectedNodes` has a noisy log line now based on my testing. The change in behavior is from the fact that we call `updateExpectedNodes` on every query instead of just on filtered queries. I may simply return to the old behavior.

I plan on fixing this once I get @mck's review so I can batch any other changes needing to be made.

michaelsembwever commented on src/java/org/apache/cassandra/index/sai/disk/v1/vector/PrimaryKeyWithScore.java:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * A PrimaryKey with one piece of metadata. Subclasses define the metadata, and to prevent unnecessary boxing, the
+ * metadata is not referenced in this calss. The metadata is not used to determine equality or hash code, but it is used
```
```suggestion
 * metadata is not referenced in this class. The metadata is not used to determine equality or hash code, but it is used
```

michaelsembwever commented on src/java/org/apache/cassandra/index/sai/disk/v1/V1SSTableIndex.java:
```diff
@@ -174,14 +175,28 @@ public List<KeyRangeIterator> search(Expression expression,
+            {
+                iterators.add(segment.orderBy(orderer, keyRange, context));
+            }
+        }
```
entirely optional
```suggestion
        for (Segment segment : segments)
            if (segment.intersects(keyRange))
                iterators.add(segment.orderBy(orderer, keyRange, context));

```

michaeljmarshall committed:
CASSANDRA-20086: Use score ordered iterators for ANN search

Rewrite SAI's ANN search query execution logic to
more efficiently merge graph search results using
similarity score-ordered (descending) iterators to
merge seg efficiently, allowing for reduce memory
consumption during queries, reduced impact of
overwrites and tombstones, selective re-querying of
minimally necessary graphs, and reduced shuffling
of Primary Key objects.

Also increases test coverage via new parameterization
logic to cover graph search and brute force code
paths.

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/23ec1c8a3ff6f34882981df4a6fd07e0e27feafc
